File: 김진효/week11/43162_네트워크.py
 # [성능 요약] 메모리: 11.5 MB 시간: 0.77 ms 

from collections import deque

# 굳이 인접 배열말고 0번 정점으로 해서
# 입력케이스 (1,1,0) 이대로 돌아보면 될 것 같은데
def solution(n, computers):
    answer = n

    visited = [False] * n
    visited[0] = True

    q = deque([i for i in range(n)])
    while q:
        now = q.popleft()
        visited[now] = True
        for i in range(n):
            if now == i:
                continue
            # 연결되어 있고 방문하지 않았다면
            if computers[now][i] == 1 and not visited[i]:
                visited[i] = True
                q.appendleft(i)
                answer -=1

    return answer

print(solution(3, [[1, 1, 0], [1, 1, 0], [0, 0, 1]]))
print(solution(3,[[1, 1, 0], [1, 1, 1], [0, 1, 1]]))
print(solution(4,[[1, 1, 0, 0], [1, 1, 0, 0], [0, 0, 1, 1], [0, 0, 1, 1]]))

File: 백태균/week11/43162_네트워크.java
import java.util.*;

class Solution {
    
    List<ArrayList<Integer>> graph;
    static boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        graph = new ArrayList<>();
        visited = new boolean[computers.length];
        
        for (int i=0; i<computers.length; i++) {
            graph.add(new ArrayList<>());
        }
        
        // 인접 리스트 구성
        for (int i=0; i<computers.length; i++) {
            for (int j=0; j<computers[0].length; j++) {
                // 자기 자신은 제외
                if (i == j) {
                    continue;
                }
                
                // 1이면 연결되어 있다는 거니깐 인접 리스트에 단방향으로 추가
                if (computers[i][j] == 1) {
                    graph.get(i).add(j);
                }
            }
        }
        
        // 연결된 네트워크 개수 세기
        int cnt = 0;
        for (int i=0; i<computers.length; i++) {
            // 방문하지 않았다는 것은 새로운 네트워크의 시작이라는 뜻
            if (!visited[i]) {
                dfs(i);
                cnt++;
            }
        }
        
        return cnt;
    }
    
    // 얼마나 깊이 연결되어 있는지 확인하기 위해 dfs 사용
    public void dfs(int startNode) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(startNode);
        visited[startNode] = true;
        
        while (!stack.isEmpty()) {
            int node = stack.pop();
            
            for (int nextNode : graph.get(node)) {
                if (!visited[nextNode]) {
                    stack.push(nextNode);
                    visited[nextNode] = true;
                }
            }
        }
    }
}